### PR TITLE
ENH: improve the speed of numpy.where using a branchless code

### PR DESCRIPTION
**Pull request by zephyr111**
Hello,

This PR speed up the `numpy.where` function using a branch-less implementation when the sparsity is >0.1. It is especially efficient on unpredictable input data because the previous implementation based on a conditional was causing many stalls due to branch missprediction. Such a case happens on random input data and often happens on real-world data. The performance of the branch-less implementation can change from one architecture from another. I do not expect it to be slower than the previous one on all modern x86-64 processors. It might be interesting to check if it is sometimes a bit slower on ARM/POWER architectures.

Here is a benchmark (only bench the modified path of the new code since the rest is left unchanged):

```python
size = 1024 * 1024 // 8
a = np.random.rand(size)
rnd_02 = a > 0.2
rnd_03 = a > 0.3
rnd_04 = a > 0.4
rnd_05 = a > 0.5
all_ones = np.ones(size, dtype=bool)
rep_zeros_2 = np.arange(size) % 2 == 0
rep_zeros_4 = np.arange(size) % 4 == 0
rep_zeros_8 = np.arange(size) % 8 == 0
rep_ones_2 = np.arange(size) % 2 > 0
rep_ones_4 = np.arange(size) % 4 > 0
rep_ones_8 = np.arange(size) % 8 > 0

test_cases = [
    rnd_02, rnd_03, rnd_04, rnd_05, 
    all_ones, 
    rep_zeros_2, rep_zeros_4, rep_zeros_8, 
    rep_ones_2, rep_ones_4, rep_ones_8
]

for test_case in test_cases:
    %timeit -n 1000 np.nonzero(test_case)

for test_case in test_cases:
    assert np.allclose(np.nonzero(test_case)[0], np.arange(size)[test_case])
```

and here is the results on my Intel i5-9600KF processor:

```python
----- OLD -----

213 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
295 µs ± 4.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
375 µs ± 5.09 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
408 µs ± 731 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
63.7 µs ± 156 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
83.6 µs ± 229 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
78.7 µs ± 1.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
70.7 µs ± 577 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
85.2 µs ± 988 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
91.9 µs ± 93.9 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
77.9 µs ± 366 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

----- NEW -----

65.5 µs ± 696 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.7 µs ± 198 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.7 µs ± 181 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.1 µs ± 138 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.5 µs ± 155 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.1 µs ± 205 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
64.8 µs ± 271 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
64.3 µs ± 353 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.4 µs ± 474 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.4 µs ± 437 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
65.7 µs ± 247 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

As you can see, the new timings are nearly always better on this machine. The new code is up to *6.3 times faster*.

### CODE REVIEW

**Review by tylerjereddy — COMMENTED** [review]
At first glance, the diff is pretty small and the C code coverage is 100 %, which is a good sign.

Since you mention other architectures as an unknown, comparing a POWER9 node on `main` vs. this branch with your script using `gcc 9.4.0`, Python `3.9.9` results in a mixture of slowdowns and speedups.

- `main`:

```
391 µs ± 633 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
515 µs ± 235 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
594 µs ± 76.5 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
620 µs ± 302 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
104 µs ± 18.3 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
600 µs ± 472 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
385 µs ± 64 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
238 µs ± 85.1 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
521 µs ± 934 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
391 µs ± 241 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
236 µs ± 45 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

- this branch:

```
398 µs ± 665 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
534 µs ± 262 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
634 µs ± 40.4 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
673 µs ± 264 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
113 µs ± 37.1 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
229 µs ± 120 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
180 µs ± 557 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
164 µs ± 717 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
231 µs ± 492 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
164 µs ± 35.1 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
168 µs ± 94.6 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

I suppose a preprocessor directive could be used to guard some architectures, though not sure how interested the core team would be in the maintenance tradeoffs there.

My other suggestion, is that we tend to use `asv` (airspeed velocity) for our standard benchmarks, and we already have some for `np.where` in `benchmarks/bench_function_base.py`. If the team is interested in the direction of these changes then:

- probably want to run `asv continuous main faster-where -b "Where"` to check our current benchmark results for these changes
- if the current benchmarks are not sensitive to the adjustments here, maybe expand them to include the appropriate data structures/sparsity requirements?
  - also, some of the slowdowns I see above may not be "statistically significant" when looking at the "errors," and `asv continuous` does a good job of summarizing the "significant" differences without needing manual parsing 

But probably best to wait to hear from the team before going into that.

**Comment by zephyr111** [comment]
Thank you for the results and the suggestions. 

The results are quite surprising because a branch-less code should not be impacted by the content of the input array. Thus, the expected run time should be relatively constant like on x86-64. Consequently, I expect GCC to produce an assembly code with conditional jumps despite the code is branch-less. However, this part of the code appear to generate a good code on the three main kind of platform based on [GodBolt results](https://godbolt.org/z/Ej1Won7rG) (using GCC 11.2 with `-O2`). Alternatively, maybe a complex low-level effect is happening POWER processors that I am not aware of. I can try to tune/optimize the code, but the main problem is I do not have access to any POWER processor (nor ARM yet) to check the performances (or at least the produced assembly code).

Regarding airspeed velocity, I am not very familiar with this tool. I my understanding is correct, it is a tool to automate the analysis of the performance of packages like Numpy over time on many platforms. I found [this page](https://pv.github.io/numpy-bench/#/summarylist) showing the results for Numpy but the current platforms are only Intel ones so far.

**Comment by seberg** [comment]
It would be awesome if you could add a few additional benchmarks (based on the ones you got) here: https://github.com/numpy/numpy/blob/61dec07ba52c0aafe1149e43bb22ffd0cb4b06c9/benchmarks/benchmarks/bench_function_base.py#L282

@tylerjereddy your benchmarks are a bit mixed, but it looks a lot like a general win anyway?  The best thing would probably to expand those `asv` benchmarks I linked, then Tyler can run them locally if we want to be sure that it is a win on POWER as well.

**Comment by tylerjereddy** [comment]
Yeah, I can run the actual `asv` benchmarks on POWER9 and ARMv8 when ready, just ping me again. Yeah, the results were mixed--probably a net win, but easier to see with `asv` instead of manual inspection of course.

**Comment by zephyr111** [comment]
I have optimized the implementation (using loop unrolling to improve the instruction level parallelism, especially on architecture with higher-latency instructions like on most ARM architectures). I also tested this on an ARM Graviton2 machine (Amazon EC2 t4g.small with a default Ubuntu 20 LTS system). Results are also pretty good:

```none
-- OLD --

259 µs ± 1.09 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
350 µs ± 555 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
444 µs ± 1.17 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
502 µs ± 779 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
115 µs ± 700 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
193 µs ± 182 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
153 µs ± 68.5 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
133 µs ± 113 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
193 µs ± 179 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
127 µs ± 203 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
115 µs ± 415 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

-- NEW --
124 µs ± 286 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
126 µs ± 1.28 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
127 µs ± 638 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
126 µs ± 1.33 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
126 µs ± 1.26 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
126 µs ± 1.08 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
124 µs ± 491 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
128 µs ± 965 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
127 µs ± 1.52 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
126 µs ± 736 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
128 µs ± 1.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

This is almost always faster and up to *4 times faster*. Results are also stable like on an x86-64 PC.

I also updated the benchmark and tested `asv` which is quite interesting and successfully report the biggest differences.

@tylerjereddy Can you run `asv` on POWER9 on this updated branch?

**Comment by rossbar** [comment]
Just a quick suggestion re: benchmarking - the `asv compare` command can be used to directly compare two commits and the results are presented in a more digestible way. The `runtests.py` script also exposes this functionality via something like: `python runtests.py --bench-compare main faster-where bench_function_base`.

More info available here: https://numpy.org/devdocs/benchmarking.html

File: benchmarks/benchmarks/bench_function_base.py [hunk 1]
```diff
@@ -293,3 +308,43 @@ def time_2(self):
...
+        np.where(self.rep_zeros_8)
+
+    def time_interleaved_ones_x2(self):
+        np.where(self.rep_ones_2)
+
+    def time_interleaved_ones_x4(self):
+        np.where(self.rep_ones_4)
+
+    def time_interleaved_ones_x8(self):
+        np.where(self.rep_ones_8)
```
Comment (seberg) [inline benchmarks/benchmarks/bench_function_base.py:349]: Thanks for those thorough tests!  We could group the `percent` and `interleaved` ones with parameters in `asv` probably, although I am not sure if that doesn't require splitting it into multiple classes.  So fine with me, just wanted to mention it.

**Review by seberg — APPROVED** [review]
I would love to not need that manual unroll there, but considering the speed-up, and that I am not sure that `NPY_GCC_OPT_3` will do the trick due to currently being limited to GCC, happy to go ahead with it.

**Comment by zephyr111** [comment]
> I would love to not need that manual unroll there, but considering the speed-up, and that I am not sure that NPY_GCC_OPT_3 will do the trick due to currently being limited to GCC, happy to go ahead with it.

Me too. I tried with a loop using default compiler parameters and GCC was generating a very bad code with it. It seems that using -O3 is sufficient to generate an efficient code similar to the unrolled loop based on [GodBolt results](https://godbolt.org/z/ovcdM4TYd). I will this in the Numpy code.

I also found out that GCC only generate a branchless code since GCC 11. Before this version, the compiler was visibly not able to generate a branchless code. Clang was able to do that since a very long time (at least since Clang 3.0 released 10 years ago). This certainly explain why @tylerjereddy got mitigated results with GCC 9.4 (that was still a bit faster) and why the results indicated that the execution was not branchless. There was open issues submitted in the GCC bug tracker to discuss about this problem in the past (like [this one](https://gcc.gnu.org/bugzilla//show_bug.cgi?id=97734) for example).

File: numpy/core/src/multiarray/item_selection.c [hunk 2]
```diff
@@ -2641,13 +2641,32 @@ PyArray_Nonzero(PyArrayObject *self)
...
+             */
             else {
-                npy_intp j;
-                for (j = 0; j < count; ++j) {
-                    if (*data != 0) {
-                        *multi_index++ = j;
-                    }
+                npy_intp * multi_index_end = multi_index + nonzero_count;
+                npy_intp j = 0;
+
```
Comment (mattip) [inline numpy/core/src/multiarray/item_selection.c:2651]:
```suggestion

                /* Manually unroll for GCC an maybe other compilers */
```

**Review by mattip — COMMENTED** [review]
LGTM. Just one nit about adding a comment to note the unrolling is on purpose

**Comment by charris** [comment]
Thanks @zephyr111 .

**Pull request merged**